[Codegen] Support scalability when hoisting (and checking) allocations

MacDue opened the pull request:
This makes use of recent work (https://github.com/llvm/llvm-project/pull/83876) to support scalability in value bounds analysis. With these changes scalable allocations (i.e. allocations that scale with vscale), can now be hoisted and pass stack limit checking.

Hoisting scalable allocations requires knowledge of the range of vscale, which is taken from the target (using the maximum range defined by the ISA). The stack limit checking just assumes `vscale = 1` (by default), since this avoids needing to look at the target (and always resolves to a constant bound), and still prevents excessive allocations.

ci-extra: build_test_all_arm64

MacDue committed:
[Codegen] Support scalability when hoisting (and checking) allocations

This makes use of recent work (https://github.com/llvm/llvm-project/pull/83876)
to support scalability in value bounds analysis. With these changes
scalable allocations (i.e. allocations that scale with vscale), can
now be hoisted and pass stack limit checking.

Hoisting scalable allocations requires knowledge of the range of vscale,
which is taken from the target (using the maximum range defined by the
ISA). The stack limit checking just assumes `vscale = 1`, since this
avoids needing to look at the target (and always resolves to a constant
bound), and still prevents excessive allocations.

banach-space commented on compiler/src/iree/compiler/Codegen/Common/Passes.td:
```diff
@@ -312,6 +312,17 @@ def HoistStaticallyBoundAllocations :
     InterfacePass<"iree-hoist-statically-bound-allocations", "mlir::FunctionOpInterface"> {
   let summary = "Hoist statically bound alloca ops to the entry block of functions";
   let constructor = "mlir::iree_compiler::createHoistStaticallyBoundAllocationsPass()";
+  let options = [
+    Option<"vscaleMin", "vscale-min", "unsigned",
+            /*default=*/"0",
+           "Minimum possible value of vscale.">,
+    Option<"vscaleMax", "vscale-max", "unsigned",
+            /*default=*/"0",
+           "Maximum possible value of vscale (a value of zero means unbounded).">
+  ];
```
1. What's the meaning of these in the case of fixed-width vectors? Could you add a comment?
2. Using `0` for `unbounded` (which happens to be the default) is a bit counter-intuitive for me. Would it be possible to test such case?

MacDue replied:
> Using 0 for unbounded (which happens to be the default) is a bit counter-intuitive for me. Would it be possible to test such case?

It's the pretty much the same as vscale_range is specified (in the LangRef). Unbounded is the default behaviour because that's saying nothing is known about vscale.

banach-space commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -121,44 +124,96 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
   }
 
+  Value vscale = nullptr;
+  if (!vscaleRange.has_value()) {
+    auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(funcOp);
+    vscaleRange = getDefaultVscaleRange(targetAttr);
+  }
```
This logic means that the value specified by the user (when configuring `HoistStaticallyBoundAllocations`) will take precedence over function attributes. This makes sense to me, but please add a comment and test this behaviour :)

banach-space commented on compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:
```diff
@@ -157,3 +157,81 @@ func.func @nested_op_alloc_linalg_use(%arg0 : index) {
+//       CHECK:     %[[CAST:.+]] = memref.cast %[[SUBVIEW]]
+//       CHECK:     linalg.fill
+//  CHECK-SAME:         outs(%[[CAST]] :
+//       CHECK:   memref.dealloc %[[ALLOC:.+]] : memref<?xi32>
```
I know that this is merely adapting an existing test rather than something specific to your PR, but isn't hoisting in this case incorrect? As in, if `memref.alloc` is hoisted then shouldn't `memref.dealloc` be hoisted as well?

MacDue replied:
The `memref.dealloc` is hoisted (though this test would not catch if it wasn't).

banach-space replied:
It should be possible to update the test to capture that, right?

MacDue replied:
Yep (I was just copy/pasting the previous test case when adding this).

banach-space commented on compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:
```diff
@@ -1,4 +1,4 @@
-// RUN: iree-opt --split-input-file --pass-pipeline="builtin.module(func.func(iree-hoist-statically-bound-allocations))" %s | FileCheck %s
+// RUN: iree-opt --split-input-file --pass-pipeline="builtin.module(func.func(iree-hoist-statically-bound-allocations{vscale-min=1 vscale-max=16}))" %s | FileCheck %s
```
Perhaps use values other then default for vscale range?

MacDue replied:
Why? This is just testing hoisting allocas, so any reasonable range should work -- with the same result (this is not testing computing the scalable bounds).

banach-space commented on compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:
To me it would make sense to "duplicate" most/all tests with a `vscale` variant. Basically, if `non_entry_bb_scalable_allocas` is next to `@non_entry_bb_allocas` then it becomes pretty obvious that the former is just a "scalable" variant of the latter. Otherwise, these two tests are quite unrelated. Not sure we need to duplicate every test though 🤔

MacDue replied:
None of the logic I touched looks at users alloca/alloc being hoisted, so I don't think it's necessary to duplicate every test. I just chose a few representative examples. I'll move the tests though :+1:

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp:
```diff
@@ -58,11 +59,14 @@ checkStackAllocationSize(mlir::FunctionOpInterface funcOp) {
-      if (succeeded(ub)) {
-        allocaSize *= ub.value();
+      // Assume vscale is 1 for determining if the alloca is within the stack
+      // limit. This should always resolve to a constant bound.
```
> Assume vscale is 1 for determining if the alloca is within the stack limit.

This feels wrong :) Could you explain what's happening here? In particular, why bother with "vscale range" in various places if such an assumption is being made here?

MacDue replied:
This is for a static (i.e. at compile time) check on the number of bytes allocated in a function. We can't know what `vscale` will actually be here, so any value will be 'wrong'. However, assuming `vscale` is 1, will still catch excessive allocations (which is what I think the intention of this check is), without being overly conservative.

This does not change codegen, it's simply a check (does this function allocate too many bytes yes/no?).

banach-space replied:
Ah, now I see, thanks for the explanation!

Would it be possible to grab the actual values though? I guess that it can only come from a function attribute and/or target info? Otherwise I'm worried that this isn't conservative enough 🤔

MacDue replied:
As far as I understand it this is just a check for unreasonably large allocations, but nothing would break if at some `vscale` it happened to go over it a bit. In the size of allocations `vscale` will a small factor (with the most of the size likely from static dimensions), so I think you can get away with mostly ignoring it (and still have a reasonable check).

MacDue replied:
Assuming `vscale = 16` (on Arm) would be assuming a value of vscale that is very unlikely to actually occur.

banach-space replied:
Yes, `vscale = 16` would be super conservative. AFAIK, the widest vectors that we have today are 512 bits (`vscale = 4`):
* https://en.wikipedia.org/wiki/Fujitsu_A64FX

However, we need to make sure that we don't underestimate too optimistically. It's tricky because this is a compile-time check and we won't know the value of `vscale` until runtime. I think that `clAssumedVscaleValue` that you added is a very good compromise for now. If we need to refine this logic then what you've added here will allow us to make an informed decision. Thanks!

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/check_ir_before_llvm_conversion_not_fail_unbound.mlir:
```diff
@@ -29,3 +29,16 @@ module {
+  func.func @scalable_alloca() {
+    %c4 = arith.constant 4 : index
+    %vscale = vector.vscale
+    %c4_vscale = arith.muli %vscale, %c4 : index
```
IIUC, all allocation in this file are meant to be out-of-bounds. However, `vscale * 4` feels to small to be out-of-bounds? What qualifies as "out-of-bounds"?

banach-space commented on compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:
```diff
@@ -1177,4 +1177,15 @@ bool hasFusedLeadingOp(linalg::LinalgOp rootOp) {
+getDefaultVscaleRange(IREE::HAL::ExecutableTargetAttr targetAttr) {
+  if (isAArch64(targetAttr)) {
+    // On AArch64 the vector length will always be between 128-bit and 2048-bit.
+    // This works out as a vscale range of 1 to 16.
```
Let's add a link: https://developer.arm.com/Architectures/Scalable%20Vector%20Extensions. Also, this is referring to SVE vector lengths rather than "vectors" in general. Worth claryfying.

banach-space commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -121,44 +124,96 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
   /// bounds, and a subview of the required size to be used as a replacement.
-  SmallVector<int64_t> staticShape;
+  SmallVector<int64_t> allocShape;
+  SmallVector<Value> scalableSizes;
```
Could we have a comment explaining the different between the two?

banach-space reviewed (commented):
Thanks Ben, makes sense!

I've made a few comments/suggestions, but the overall looks good.

MacDue committed:
Review fixups

hanhanW commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.h:
```diff
@@ -52,7 +52,8 @@ template <typename AllocLikeOpType>
     MemRefType allocaType, ValueRange dynamicSizes,
-    std::optional<uint64_t> alignment);
+    std::optional<uint64_t> alignment,
+    std::optional<VscaleRange> vscaleRange = {});
```
nit: it looks better if the default value is `std::nullopt`, but not `{}`. Same for below changes

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:
```diff
@@ -1177,4 +1177,16 @@ bool hasFusedLeadingOp(linalg::LinalgOp rootOp) {
+    // On AArch64 the scalable vector length will always be between 128-bit and
+    // 2048-bit. This works out as a vscale range of 1 to 16. See:
+    // https://developer.arm.com/Architectures/Scalable%20Vector%20Extensions
+    return VscaleRange{1, 16};
```
(Every time I see this, I'd like to suggest to use designated initializers. Then I realized that it is a c++20 feature. :p)

hanhanW commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -121,44 +123,101 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
+    scalableBound = nullptr;
+    if (vscaleRange.has_value()) {
+      // Scalability supported: Allocations could be scalable.
+      auto ub = vector::ScalableValueBoundsConstraintSet::computeScalableBound(
```
Please spell the type (i.e., do not use auto) because:

1.  It is not trivial.
2. We do access a member of the variable.

hanhanW commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -121,44 +123,101 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
+      } else {
+        // Static size.
+        allocShape.push_back(constantBound);
+      }
```
I think both alloc and alloca op take the below build method. Is it possible to let `computeAllocationBound` return `std::optional<OpFoldResult>` and use it in allocation shape?

```
static void build(::mlir::OpBuilder &odsBuilder, ::mlir::OperationState &odsState, ArrayRef<OpFoldResult> sizes, Type elementType, Attribute memorySpace = {})
```

MacDue replied:
That's much nicer, thanks.

MacDue replied:
Seems like it may break everything too... :pensive:

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp:
```diff
@@ -58,11 +59,14 @@ checkStackAllocationSize(mlir::FunctionOpInterface funcOp) {
+          /*vscaleMin=*/1, /*vscaleMax=*/1, presburger::BoundType::UB);
+      FailureOr<vector::ConstantOrScalableBound::BoundSize> size;
+      if (succeeded(ub) && succeeded(size = ub->getSize())) {
+        allocaSize *= size->baseSize;
```
I don't get why we need `size`. Can't this just be below?

```
      if (succeeded(ub)) {
        allocaSize *= ub->getSize()->baseSize;
        continue;
      }
```

MacDue replied:
Probably overly cautious,  `getSize()` can fail if the bound is a more complex expression (in terms of vscale). However, that should never happen if vscale is a constant (which it is here). So `ub->getSize()->baseSize` should be fine :+1:.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp:
```diff
@@ -28,8 +28,17 @@ struct HoistStaticallyBoundAllocationsPass
+    if (this->vscaleMax == 0 || this->vscaleMin > this->vscaleMax)
+      return {};
+    return VscaleRange{this->vscaleMin, this->vscaleMax};
+  }();
```
Why do we need the lambda? I think we can simplify it to

```cpp
std::optional<VscaleRange> vscaleRange;
if (this->vscaleMax != 0 && this->vscaleMin <= this->vscaleMax) {
  vscaleRange = {this->vscaleMin, this->vscaleMax};
}
```

MacDue committed:
Review fixups (round 2)

MacDue committed:
CI fix

MacDue committed:
Add `iree-llvmcpu-stack-allocation-assumed-vscale` option

This allows changing the value of vscale used when checking if
(scalable) allocations are within the stack limit. Internal suggestion
from @ banach-space.

hanhanW commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -121,44 +123,92 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
+    SmallVector<int64_t> staticShape;
+    SmallVector<Value> dynamicSizes;
+    staticShape.reserve(allocLikeType.getRank());
+    dynamicSizes.reserve(allocLikeType.getRank());
```
[optional] I'd avoid using the trick here. This kind of tricks should be happening in `dispatchIndexOpFoldResults` implementation details.

hanhanW commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -121,44 +123,92 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
+      subviewSizes.push_back(dynamicSize);
+    }
+
+    // FIXME: The `AllocLikeOp::build()` method for OpFoldResult drops the
+    // layout, so we have to resolve the static/dynamic values here.
```
ohh.. this is unfortunate..

MaheshRavishankar replied:
Fix this upstream?

MacDue replied:
It should be :) 
But I'd prefer not to wait on the fix (and integration)

hanhanW replied:
Yes, let's do it in a follow-up..

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/Utils.h:
```diff
@@ -237,6 +237,14 @@ void sinkOpsInCFG(const SmallVector<Operation *> &allocs,
 // the inputs.
 bool hasFusedLeadingOp(linalg::LinalgOp rootOp);
 
+struct VscaleRange {
+  unsigned min;
+  unsigned max;
+};
```
Would it make sense to make it has initial values (i.e., `0`) like what we have in `Passes.td`? In this case, perhaps we can avoid `std::optional`. Instead, we can pass `VscaleRange vscaleRange = VscaleRange{}` in the methods. What do you think?

MacDue replied:
I'd prefer to keep the `std::optional`. I don't like encoding the semantics in the value, it's just the option parser does not support `std::optional<unsigned>`, so I have to do that in `Passes.td`.

hanhanW replied:
I see, makes sense to me. Thanks!

hanhanW reviewed (approved):
LGTM, just few nits

banach-space reviewed (approved):
Thanks for addressing my comments, LGTM!

MacDue committed:
Review fixups
